pkg-manager/core/test/catalogs.ts

```diff
@@ -297,6 +297,60 @@ test('lockfile catalog snapshots retain existing entries on --filter', async ()
   })
 })
 
+// Regression test for https://github.com/pnpm/pnpm/issues/8638
+test('lockfile catalog snapshots do not contain stale references on --filter', async () => {
+  const { options, projects, readLockfile } = preparePackagesAndReturnObjects([
+    {
+      name: 'project1',
+      dependencies: {},
+    },
+    {
+      name: 'project2',
+      dependencies: {
+        'is-positive': 'catalog:',
+      },
+    },
+  ])
+
+  await mutateModules(installProjects(projects), {
+    ...options,
+    lockfileOnly: true,
+    catalogs: {
+      default: {
+        'is-positive': '^1.0.0',
+      },
+    },
+  })
+
+  expect(readLockfile().catalogs).toStrictEqual({
+    default: {
+      'is-positive': { specifier: '^1.0.0', version: '1.0.0' },
+    },
+  })
+
+  // This test updates the catalog entry in project2, but only performs a
+  // filtered install on project1. The lockfile catalog snapshots for project2
+  // should still be updated despite it not being part of the filtered install.
```

Comment:
If you do this, `is-positive` in project2 will remain at v1. What will be store in the importers section? Will `is-positive` correctly be updated on a subsequent full install?

Reply:
It correctly goes to `is-positive@3.1.0`! This is thanks to your change to compute the lockfile for outdated projects before performing a filtered install. 🙂

- https://github.com/pnpm/pnpm/pull/8183

Reply:
It might be good to make sure that stays the case in this test. Let me add a check that `importers.project2` looks as expected.

New commit for that here: https://github.com/pnpm/pnpm/pull/9126/commits/c7f4d72e8f4985b87c3530c75d060f6679d58535